### PR TITLE
Add file creation

### PR DESCRIPTION
**Pull request by TheNeikos**
@matthiasbeyer, @neithernut A possible implementation of a lazy enum to have lazy loading of files.

### CODE REVIEW

**Comment by TheNeikos** [comment]
Be careful, the tests are not yet run due to #100 not being merged.

File: libimagstore/src/lazyfile.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    fn lazy_file() {
+        let path = PathBuf::from("/tmp/test");
+        let mut lf = LazyFile::new(path);
+
+        write!(lf.create_file().unwrap(), "Hello World").unwrap();
+    }
+
+    #[test]
+    fn lazy_file_with_file() {
+        let path = PathBuf::from("/tmp/test2");
```
Comment (matthiasbeyer) [inline libimagstore/src/lazyfile.rs]: Use https://crates.io/crates/tempdir/ here please.

**Comment by neithernut** [comment]
I didn't spot anything fishy yet.

File: libimagstore/src/error.rs [hunk 2]
```diff
@@ -19,6 +21,8 @@ fn store_error_type_as_str(e: &StoreErrorKind) -> &'static str {
     match e {
         &StoreErrorKind::IdNotFound  => "ID not found",
         &StoreErrorKind::OutOfMemory => "Out of Memory",
+        &StoreErrorKind::FileNotFound => "File corresponding to ID not found",
+        &StoreErrorKind::FileNotCreated => "File corresponding to ID could not be created",
```
Comment (matthiasbeyer) [inline libimagstore/src/error.rs]: Could you please align properly? So the `=>` are all in one column?

(I know I am annoying...)

File: libimagstore/src/lazyfile.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        LazyFile::File(f)
+    }
+
+    pub fn get_file(&mut self) -> Result<&File, StoreError> {
+        let file = match *self {
+            LazyFile::File(ref f) => return Ok(f),
+            LazyFile::Absent(ref p) => {
+                try!(open_file(p).map_err(|e| {
+                    StoreError::new(StoreErrorKind::FileNotFound,
+                                    Some(Box::new(e)))
```
Comment (matthiasbeyer) [inline libimagstore/src/lazyfile.rs]: There is no need for putting this into two lines, is there?

File: libimagstore/src/lazyfile.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+fn open_file<A: AsRef<Path>>(p: A) -> ::std::io::Result<File> {
+    OpenOptions::new().write(true).read(true).open(p)
+}
+
+fn create_file<A: AsRef<Path>>(p: A) -> ::std::io::Result<File> {
+    OpenOptions::new().write(true).read(true).create(true).open(p)
+}
+
+impl LazyFile {
+    pub fn new(p: PathBuf) -> LazyFile {
```
Comment (matthiasbeyer) [inline libimagstore/src/lazyfile.rs]: Shall we call this `new_with_path()` for consistency with `new_with_file()`?

Also: blank lines around functions!

File: libimagstore/src/lazyfile.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                                    Some(Box::new(e)))
+                }))
+
+            }
+        };
+        *self = LazyFile::File(file);
+        if let LazyFile::File(ref f) = *self {
+            return Ok(f);
+        }
+        unreachable!()
```
Comment (matthiasbeyer) [inline libimagstore/src/lazyfile.rs]: Could one possibly implement this function with `get_file_mut()` and `map()` over the result to remove the `mut` from the `&mut File`? Just curious...

**Comment by matthiasbeyer** [comment]
Care to add some documentation on the functions?

File: libimagstore/src/lazyfile.rs (hunk 4 above)
Comment (TheNeikos) [inline libimagstore/src/lazyfile.rs]: The idea is that you have the 'intended way' and an alternative.

File: libimagstore/src/lazyfile.rs (hunk 5 above)
Comment (TheNeikos) [inline libimagstore/src/lazyfile.rs]: That should be possible, you could also just reborrow to remove the mut :) Will do that!

File: libimagstore/src/lazyfile.rs (hunk 4 above)
Comment (matthiasbeyer) [inline libimagstore/src/lazyfile.rs]: ACK.

**Comment by TheNeikos** [comment]
If someone can explain why that is not working that would be great

**Comment by matthiasbeyer** [comment]
@TheNeikos ff-merge my "fix-pr-115" branch

**Comment by matthiasbeyer** [comment]
@TheNeikos Feel also free to squash my change into your commit (feel free to add `Suggested-by:`-Tag in the commit message)!

**Comment by TheNeikos** [comment]
Why is this blocked?

**Comment by matthiasbeyer** [comment]
Because, afaik, we rework the interface of the store, which affects this PR, doesn't it?

**Comment by TheNeikos** [comment]
No, as this is merely an abstraction below the naming of the file.

**Comment by matthiasbeyer** [comment]
@TheNeikos is this ready for merge then?

**Comment by TheNeikos** [comment]
I think so yes

**Comment by matthiasbeyer** [comment]
@homu r+

**Comment by TheNeikos** [comment]
Home did not merge? I think she fucked something up here....

**Comment by matthiasbeyer** [comment]
@homu r+

**Comment by matthiasbeyer** [comment]
@homu retry

**Comment by matthiasbeyer** [comment]
Hm. Homu is still not merging. Something seems to be broken. @TheNeikos What shall we do?

**Pull request merged**